### ISSUE TITLE
Should the order of custom sections be specified?

### ISSUE THREAD

**Issue by NWilson** [WebAssembly/tool-conventions#45]
Currently, the Linking.md spec defines two new custom sections, `"linking"` and `"reloc.*"`. However, their order is not specified with relation to each other or other sections.

Currently, the order of all the standard Wasm sections and the `"name"` section is precisely specified, with no flexibility.

I believe that whenever new Wasm custom sections are speced, their order should be included in the spec. This will ensure that sections are created in one and only one order, ensuring compatibility between tools. If two different specs define their own custom sections, there is the open question of how those specs would reference each other to define the relative ordering of their new sections. However, we aren't there yet - all I'm concerned about right now is whether the new Linking.md sections should have their order specified relative to the existing standard sections.

The sections should have this order:

| Section id/name | Rationale |
| ----------- | -----------|
| ... | ... |
| 10 | CODE section: core spec |
| 11 | DATA section: core spec |
| `"linking"` | Linking.md: MUST come after DATA since the number of segments must be known in order to validate the symbol table. This custom section must be unique. |
| `"reloc.*"` | Linking.md: MUST come after `"linking"` since the symbol table must be known in order to validate the reloc indexes. If there is more than one reloc section, they way appear in any order. |
| `"name"` | Core spec says this MUST come after DATA. We _could_ leave its order relative to linking/reloc unspecified, or could nail it down to come after them (preferable to me, to ensure compat). |

**Comment by NWilson** [comment]
Upvote previous comment to agree with nailing down the order proposed.

Upvote this comment to vote for "leave ordering between "name" section and Linking sections flexible".

Comment otherwise.

**Comment by sbc100** [comment]
I'm in favor of specifying the order of the linking metadata sections in Linking.md relative to each other and relative to the builtin section types.

I don't see why we would want to specify the order relative to other custom sections (such as "name" or any other custom sections).  Unless they do depend on each other of course.  I don't think that is the case for "name".

**Comment by dschuff** [comment]
In the context of  https://reviews.llvm.org/D54924 (and also https://reviews.llvm.org/D44024) this discussion came up again.
A binary could have wasm-specified sections, "known" custom sections (i.e. those covered by the tool-conventions doc, e.g. linking and relocation sections, producer section) as well as "unknown" custom sections (anything else).

I think we do want LLVM to check any ordering for known sections that the spec requires (including for the new event section, which now has the property that the sections numbers aren't monotonic).

So then I think the only question remaining is do we want to specify anything about the placement of the known custom sections?

1. IIRC the spec allows custom sections anywhere (even before or between spec-defined sections) so we (i.e. the tool conventions spec and LLVM) have to be prepared to at least ignore any unknown custom section anywhere in the binary. in other words we can't be prescriptive about the placement of unknown custom sections.
1a. A corollary of that is that even for our known custom sections, we shouldn't be globally prescriptive (i.e. we shouldn't say "this section must be the very first section in the file, or first one after the known sections) because if 2 different conventions each say that, then they become incompatible.

I think the current dynamic linking convention violates 1a, so maybe we should fix that.

Other than that I guess I agreed with @sbc100 in March that we should specify the relative order of the known custom sections. That still seems good to me, although I don't have a strong opinion about it.

**Comment by dschuff** [comment]
(/cc @aheejin )

**Comment by aheejin** [comment]
To enable checking of DAG-style relationship between sections, for example, both section B and C should be after section A but B and C does not have ordering with respect to each other, we should keep the history of read sections while reading. Currently both https://reviews.llvm.org/D44024 and https://reviews.llvm.org/D54924 only checks ordering between adjacent sections and bail out in case of unknown sections. The restriction of this is we have to specify one specific ordering at least for all known sections we want to check and can't allow DAG-style checking. So in the example above, we end up only accepting A-B-C and not A-C-B, while both orderings should be fine.

**Comment by binji** [comment]
For simplicity, I think all sections defined here should be specified in a total order. For the `reloc.*` sections, they  should follow the same order of their known section counterpart. We could keep the ordering looser than that, but what is the benefit?

**Comment by aheejin** [comment]
I think so far people agreed on that we should check both regular sections and known custom sections. And also it wouldn't hurt to give a total ordering for all known (regular + custom) sections. Do people have any other opinions?

And @NWilson, would you like to revive https://reviews.llvm.org/D44024, modify it to match the current code, and land it? Because he's not in active development these days I'm not sure. Otherwise I can modify https://reviews.llvm.org/D54924 to include custom section checking.

**Comment by sbc100** [comment]
I tend to agree with heejin that total ordering might make is more strict that the spec.  We should do exactly what the specs say.  If, for example, the `name` and `reloc` sections don't mention their relative order in the specs we should not impose an arbitrary one.

**Comment by sbc100** [comment]
For example https://github.com/WebAssembly/tool-conventions/blob/master/ProducersSection.md.  There is no reason to define where it should appear relative to the linking section is there?

**Comment by binji** [comment]
Certainly other tools can put custom sections wherever they want. But as long as we're specifying them here, I think we should be more strict, at least at first. I think we should do it for the `producers` section too. If we don't, I think it's likely that the first producer that produces this section will de facto specify where it belongs. It's more work for consumers too, since they'll need to test that each custom section works properly in all the various locations.

**Comment by sbc100** [comment]
It just seems strange to ask authors of independent spec to specify the relative ordering of their unrelated custom sections.

And it would also be strange for llvm to impose and ordering that wasn't specified in said specs.

Currently we don't have many so its not a huge deal I suppose.

**Comment by aheejin** [comment]
I now think it may be better not to impose a total ordering unless we strictly need to. When we generate all the (regular + custom) known sections not in random order but a certain predetermined order, but the place we check the ordering is not the object file writer but the reader. Wouldn't it be weird if our LLVM tools, let's say, `obj2yaml`, or lld `wasm-ld`, can't even read valid object files generated?

**Comment by aheejin** [comment]
Added custom order specification to the Linking spec in #84.
